apiserver/handler.go

```diff
@@ -0,0 +1,230 @@
+package apiserver
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"istio.io/manager/model"
+
+	restful "github.com/emicklei/go-restful"
+	"github.com/golang/glog"
+)
+
+var schema model.ProtoSchema
+
+// GetConfig retrieves the config object from the configuration registry
+func (api *API) GetConfig(request *restful.Request, response *restful.Response) {
+
+	params := request.PathParameters()
+	key, err := setup(params)
+	if err != nil {
+		api.writeError(http.StatusBadRequest, err.Error(), response)
+		return
+	}
+
+	proto, ok := api.registry.Get(key)
+	if !ok {
+		errLocal := &model.ItemNotFoundError{Key: key}
+		api.writeError(http.StatusNotFound, errLocal.Error(), response)
+		return
+	}
+
+	retrieved, err := schema.ToJSON(proto)
+	if err != nil {
+		api.writeError(http.StatusInternalServerError, err.Error(), response)
+		return
+	}
+	var retJSON interface{}
+	err = json.Unmarshal([]byte(retrieved), &retJSON)
+	if err != nil {
+		api.writeError(http.StatusInternalServerError, err.Error(), response)
+		return
+	}
+	config := Config{
+		Name: params["name"],
+		Type: params["kind"],
+		Spec: retJSON,
+	}
+	if err = response.WriteHeaderAndEntity(http.StatusOK, config); err != nil {
+		api.writeError(http.StatusInternalServerError, err.Error(), response)
+	}
+}
+
+// AddConfig creates and stores the passed config object in the configuration registry
+// It is equivalent to a restful PUT and is idempotent
+func (api *API) AddConfig(request *restful.Request, response *restful.Response) {
+
+	// ToDo: check url name matches body name
+
+	params := request.PathParameters()
+	key, err := setup(params)
+	if err != nil {
+		api.writeError(http.StatusBadRequest, err.Error(), response)
+		return
+	}
+
+	config := &Config{}
+	err = request.ReadEntity(config)
+	if err != nil {
+		api.writeError(http.StatusBadRequest, err.Error(), response)
+		return
+	}
+
+	err = config.ParseSpec()
+	if err != nil {
+		api.writeError(http.StatusBadRequest, err.Error(), response)
+		return
+	}
+
+	err = api.registry.Post(key, config.ParsedSpec)
+	if err != nil {
+		response.AddHeader("Content-Type", "text/plain")
+		switch err.(type) {
+		case *model.ItemAlreadyExistsError:
+			api.writeError(http.StatusConflict, err.Error(), response)
+			return
```

Comment:
(nit) - The multiple `returns` can be moved outside of the switch statement since this error path returns regardless of the error type.